File: src/main/java/org/wysko/midis2jam2/instrument/family/percussion/drumset/HiHat.java
/*
 * Copyright (C) 2021 Jacob Wysko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.wysko.midis2jam2.instrument.family.percussion.drumset;

import com.jme3.math.Quaternion;
import com.jme3.math.Vector3f;
import com.jme3.scene.Node;
import com.jme3.scene.Spatial;
import org.wysko.midis2jam2.Midis2jam2;
import org.wysko.midis2jam2.instrument.family.percussion.CymbalAnimator;
import org.wysko.midis2jam2.midi.MidiNoteOnEvent;

import java.util.List;
import java.util.stream.Collectors;

import static org.wysko.midis2jam2.Midis2jam2.rad;

/**
 * The hi-hat.
 */
public class HiHat extends SingleStickInstrument {
	
	private final static int WOBBLE_SPEED = 10;
	
	private final static double DAMPENING = 2;
	
	private final static double AMPLITUDE = 0.25;
	
	/**
	 * The list of NoteOn events that the stick needs to worry about (closed and open).
	 */
	private final List<MidiNoteOnEvent> hitsToStrike;
	
	/**
	 * The Top cymbal.
	 */
	private final Node topCymbal = new Node();
	
	/**
	 * The Whole hat.
	 */
	private final Node wholeHat = new Node();
	
	/**
	 * The cymbal animator.
	 */
	private final CymbalAnimator animator;
	
	/**
	 * The current animation time.
	 */
	private double animTime;
	
	/**
	 * The current status of the hi-hat.
	 *
	 * @see HiHatStatus
	 */
	private HiHatStatus status = HiHatStatus.CLOSED;
	
	public HiHat(Midis2jam2 context, List<MidiNoteOnEvent> hits) {
		super(context, hits);
		hitsToStrike = hits.stream().filter(h -> h.note == 42 || h.note == 46).collect(Collectors.toList());
		Spatial topCymbalModel = context.loadModel("DrumSet_Cymbal.obj", "CymbalSkinSphereMap.bmp",
				Midis2jam2.MatType.REFLECTIVE, 0.7f);
		Spatial bottomCymbalModel = context.loadModel("DrumSet_Cymbal.obj", "CymbalSkinSphereMap.bmp",
				Midis2jam2.MatType.REFLECTIVE, 0.7f);
		bottomCymbalModel.setLocalRotation(new Quaternion().fromAngles(rad(180), 0, 0));
		
		topCymbal.setLocalTranslation(0, 1.2f, 0);
		topCymbal.attachChild(topCymbalModel);
		Node bottomCymbal = new Node();
		bottomCymbal.attachChild(bottomCymbalModel);
		wholeHat.attachChild(topCymbal);
		wholeHat.attachChild(bottomCymbal);
		wholeHat.setLocalScale(1.3f);
		
		wholeHat.setLocalTranslation(0, 0, -14);
		highLevelNode.attachChild(wholeHat);
		highLevelNode.setLocalTranslation(-6, 22, -72);
		highLevelNode.setLocalRotation(new Quaternion().fromAngles(0, rad(90), 0));
		highLevelNode.detachChild(stickNode);
		wholeHat.attachChild(stickNode);
		stickNode.setLocalTranslation(0, 1, 13);
		this.animator = new CymbalAnimator(AMPLITUDE, WOBBLE_SPEED, DAMPENING);
	}
	
	@Override
	public void tick(double time, float delta) {
		animator.tick(delta);
		MidiNoteOnEvent recoil = null;
		while (!hits.isEmpty() && context.file.eventInSeconds(hits.get(0)) <= time) {
			recoil = hits.remove(0);
		}
		
		if (recoil != null) {
			animator.strike();
			wholeHat.setLocalTranslation(0, (float) (-0.7 * velocityRecoilDampening(recoil.velocity)), -14);
			if (recoil.note == 46) {
				status = HiHatStatus.OPEN;
				topCymbal.setLocalTranslation(0, 2, 0);
			} else {
				status = HiHatStatus.CLOSED;
				topCymbal.setLocalTranslation(0, 1.2f, 0);
			}
			animTime = 0;
		}
		topCymbal.setLocalRotation(new Quaternion().fromAngles(status == HiHatStatus.CLOSED ? 0 : animator.rotationAmount(), 0, 0));
		if (animTime != -1) animTime += delta;
		handleStick(time, delta, hitsToStrike);
		
		wholeHat.move(0, 0.025f, 0);
		if (wholeHat.getLocalTranslation().y > 0) {
			Vector3f localTranslation = new Vector3f(wholeHat.getLocalTranslation());
			localTranslation.y = Math.min(localTranslation.y, 0);
			wholeHat.setLocalTranslation(localTranslation);
		}
	}
	
	/**
	 * The status of the hi-hat.
	 */
	private enum HiHatStatus {
		/**
		 * The hat is closed, meaning the top cymbal and bottom cymbal are together.
		 */
		CLOSED,
		
		/**
		 * The hat is open, meaning the top cymbal is raised.
		 */
		OPEN
	}
}

File: src/main/java/org/wysko/midis2jam2/instrument/family/percussion/drumset/HiHat.kt
/*
 * Copyright (C) 2021 Jacob Wysko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */
package org.wysko.midis2jam2.instrument.family.percussion.drumset

import com.jme3.math.Quaternion
import com.jme3.math.Vector3f
import com.jme3.scene.Node
import org.wysko.midis2jam2.Midis2jam2
import org.wysko.midis2jam2.instrument.algorithmic.NoteQueue
import org.wysko.midis2jam2.instrument.family.percussion.CymbalAnimator
import org.wysko.midis2jam2.midi.Midi
import org.wysko.midis2jam2.midi.MidiNoteOnEvent
import org.wysko.midis2jam2.util.MatType
import org.wysko.midis2jam2.util.Utils.rad

/** The hi-hat. */
class HiHat(context: Midis2jam2, hits: MutableList<MidiNoteOnEvent>) : SingleStickInstrument(context, hits) {

    /** The list of NoteOn events that the stick needs to worry about (closed and open). */
    private val hitsToStrike: List<MidiNoteOnEvent>

    /** The top cymbal. */
    private val topCymbal = Node()

    /** The whole hi-hat. */
    private val wholeHat = Node()

    /** The cymbal animator. */
    private val animator: CymbalAnimator

    /** The current animation time. */
    private var animTime = 0.0

    /**
     * The current status of the hi-hat.
     *
     * @see HiHatStatus
     */
    private var status = HiHatStatus.CLOSED

    override fun tick(time: Double, delta: Float) {
        animator.tick(delta)

        val recoil = NoteQueue.collectOne(hits, context, time)

        /* If a note is to be played */
        if (recoil != null) {
            /* Strike for cymbal animation */
            animator.strike()

            /* Recoil hi-hat */
            wholeHat.setLocalTranslation(0f, (-0.7 * velocityRecoilDampening(recoil.velocity)).toFloat(), -14f)

            /* Open or close hat based on note */
            if (recoil.note == Midi.OPEN_HI_HAT) {
                status = HiHatStatus.OPEN
                topCymbal.setLocalTranslation(0f, 2f, 0f)
            } else {
                status = HiHatStatus.CLOSED
                topCymbal.setLocalTranslation(0f, 1.2f, 0f)
            }

            /* Reset animation time */
            animTime = 0.0
        }
        /* Apply wobble if the hat is open */
        topCymbal.localRotation =
            Quaternion().fromAngles(if (status == HiHatStatus.CLOSED) 0f else animator.rotationAmount(), 0f, 0f)

        /* Increment anim time if there has already been a hat strike already */
        if (animTime != -1.0) animTime += delta.toDouble()

        /* Animate stick */
        handleStick(time, delta, hitsToStrike as MutableList<MidiNoteOnEvent>)

        /* Move the hat up for recoil */
        wholeHat.move(0f, 5 * delta, 0f)

        /* But not so far that it keeps going up */
        if (wholeHat.localTranslation.y > 0) {
            val localTranslation = Vector3f(wholeHat.localTranslation)
            localTranslation.also { it.y = it.y.coerceAtMost(0f) }
            wholeHat.localTranslation = localTranslation
        }
    }

    /** The status of the hi-hat. */
    private enum class HiHatStatus {
        /** The hat is closed, meaning the top cymbal and bottom cymbal are together. */
        CLOSED,

        /** The hat is open, meaning the top cymbal is raised. */
        OPEN
    }

    companion object {
        /** How fast the hi-hat wobbles. */
        private const val WOBBLE_SPEED = 10

        /** How fast the hi-hat returns to rest after being struck. */
        private const val DAMPENING = 2.0

        /** The intensity of the strike. */
        private const val AMPLITUDE = 0.25
    }

    init {
        val bottomCymbal = Node()

        /* Filter out hits that the stick needs to worry about */
        hitsToStrike = hits.filter { it.note == Midi.OPEN_HI_HAT || it.note == Midi.CLOSED_HI_HAT }

        /* Load the cymbals */
        topCymbal.attachChild(
            context.loadModel(
                "DrumSet_Cymbal.obj",
                "CymbalSkinSphereMap.bmp",
                MatType.REFLECTIVE,
                0.7f
            )
        )
        bottomCymbal.attachChild(
            context.loadModel(
                "DrumSet_Cymbal.obj",
                "CymbalSkinSphereMap.bmp",
                MatType.REFLECTIVE,
                0.7f
            ).apply {
                localRotation = Quaternion().fromAngles(rad(180.0), 0f, 0f)
            }
        )

        topCymbal.setLocalTranslation(0f, 1.2f, 0f)

        wholeHat.run {
            attachChild(topCymbal)
            attachChild(bottomCymbal)
            setLocalScale(1.3f)
            setLocalTranslation(0f, 0f, -14f)
        }

        highLevelNode.run {
            attachChild(wholeHat)
            setLocalTranslation(-6f, 22f, -72f)
            localRotation = Quaternion().fromAngles(0f, rad(90.0), 0f)
            detachChild(stickNode)
        }

        wholeHat.attachChild(stickNode)
        stickNode.setLocalTranslation(0f, 1f, 13f)
        animator = CymbalAnimator(AMPLITUDE, WOBBLE_SPEED.toDouble(), DAMPENING)
    }
}